### DISCUSSION TITLE
How to publish apk with ARMv8 Adobe Animate CC 2017?

### DISCUSSION THREAD

**Discussion by hadisn** [airsdk/Adobe-Runtime-Support#746] [Q&A]
Can someone please explain how to publish apk with 64bit support. There is no option in Adobe Animate CC 2017, what should i do? Thanks

**Comment by ajwfrost** [comment]
To override the default packaging from Animate, you need to create a file `air.cfg` and put this into the SDK's "lib" folder, i.e. the same folder as `adt.jar`. Contents should be:
```
DefaultArch=armv8
OverrideArch=armv8
DebugOut=false
```

Then if you package again ADT should override the option provided by Animate (armv7) and instead use the ARMv8 binary..

**Comment by hadisn** [comment]
Thank you for answer, but is there option to publish both ARMv8 anf ARMv7?

**Comment by ajwfrost** [comment]
Not into the same APK, each architecture ARMv7/x86/ARMv8 is packaged into a separate APK file with different version codes but the same application, which allows these to be uploaded to the Play Store where the appropriate one will be selected for installation onto an end user's handset.

thanks

**Comment by hadisn** [comment]
I always get only armeabi-v7a i can't get ARMv8...
![Capture](https://user-images.githubusercontent.com/2553775/61170842-40e1bf00-a56f-11e9-8a76-514642d89186.JPG)

**Comment by ajwfrost** [comment]
Ah, sorry, my mistake! Filename should be **adt.cfg**, rather than **air.cfg**...

**Comment by hadisn** [comment]
Got it, thanks. Now i have this error from google play console: This APK will not be served to any users because it is completely shadowed by one or more APKs with higher version codes.

Tried to upload different apks for each architecture as you adviced but i receive error that apk with lower version code will not be server to any users. 
![Capture](https://user-images.githubusercontent.com/2553775/61173492-6208d680-a594-11e9-9214-18a75497f12c.JPG)

**Comment by ajwfrost** [comment]
Interesting ... we have to have different version codes for different architectures, but it shouldn't be completely shadowed as the ABIs are different. Are you able to confirm that the different APKs have got the correct lib folders for ARMv7/ARMv8? and perhaps let us know the version codes that are within each of the APKs (via the Android Studio tool)?

**Comment by hadisn** [comment]
I had higher version code for ARMv7 and lower for ARMv8, now i changed that and ARMv8 have higher version code. I dont use Android studio tool i use Adobe Animate cc 2017 to create apks. I am uploading it for closed alpha testing and now i didn't received error as above just "Unoptimized APK" error but it should be ok with that, right?

![Capture](https://user-images.githubusercontent.com/2553775/61173668-a8f7cb80-a596-11e9-8514-2666c76a4a70.JPG)

**Comment by jeuxdaily** [comment]
hello
I put adt.cfg file in lib folder

but I compile to apk this message show

![erro_sdt_cfg](https://user-images.githubusercontent.com/14032996/63231634-6f118900-c216-11e9-84d0-b5dbe3204e2f.PNG)

**Comment by ajwfrost** [comment]
'Not a valid native extension file' implies there's something wrong with this ANE.. Milkman plugins aren't being updated for 64-bit Android targets, as far as I know..

BTW when creating multiple APKs for the different architectures then yes, the 64-bit one needs to be a higher version number than the 32-bit one. This is because 64-bit versions of Android can actually run 32-bit software so a higher version number on the 32-bit package would mean this is preferred by Play Store to the 64-bit package - hence the 64-bit package being completely overshadowed.

We've added more detailed instructions to the latest release notes going out shortly.

**Comment by Rpjstudio** [comment]
we also have the same errors and we are using milkyman ane's  so could we make a 64 bit apk or not ?
can you help me please ? we are using latest AIRSDK v 33
![1](https://user-images.githubusercontent.com/54402340/63513003-bc397780-c502-11e9-83eb-615e09946b15.PNG)

**Comment by jeuxdaily** [comment]
hi
this error come when you are update milkman extension to add 64bit
because milkman not support google play 64bit at this moment

**Comment by jeuxdaily** [comment]
Yes me too
Cause milkman doesn't update plugins 7 years ago

**Comment by newapproach** [comment]
I have also a problem. this problem is big. anyone can make a video tutorial.

 (How To Publish 64bit and 32bit Apk With Animate 2019 And Upload To Google Play Store. )

**Comment by smbln** [comment]
same problem here :(

**Comment by jeuxdaily** [comment]
Le mar. 24 sept. 2019 à 13:17, smbln <notifications@github.com> a écrit :

> same problem here :(
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/Gamua/Adobe-Runtime-Support/issues/65?email_source=notifications&email_token=ADLCAZGSTPWNHP2DGWJ5YIDQLIANNA5CNFSM4H7HGPAKYY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOD7OC3OA#issuecomment-534523320>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ADLCAZHO7YPBZ5WINAQ4AD3QLIANNANCNFSM4H7HGPAA>
> .
>
You can do it with adobe
Publish first apk 32bit then 64 bit

**Comment by P2b10** [comment]
Hi ajwfrost

I am trying to pack a 64 .apk in Adobe Animate 2015 using a 1024 certificate because that´s the one I have been using for years for my app and changing to a 2048 certificate implies presenting a new app from scratch in Google Play which would imply loosing income.

I have been using AIR SDK 33.1.1.63 and the latest AIR SDK 33.1.1.86. 

As in Adobe Animate 2015 there is no way to select ARMv7 (32 bit) or ARMv8 (64 bit) I have been trying to use the “adt.cfg” file on the SDK lib with:

this
#DefaultArch=armv8
#OverrideArch=armv8

or this
DefaultArch=armv8
OverrideArch=armv8
DebugOut=false

but nothing works. I keep getting an “armeabi-v7a” folder in the lib folder of the resulting apk.

Any solution?

Best Regards,

**Comment by ajwfrost** [comment]
@P2b10 the config file mechanism should still be in place. I’m wondering whether the file is in the right place - or some folk had issues with a file called adt.cfg but with Windows hiding known file extensions such that the file was actually adt.cfg.cfg hence we didn’t load it.

To check that you’re looking at the right location for the file, try deleting/renaming adt.jar and check that Animate then fails to package the app perhaps?

Thanks

**Comment by newapproach** [comment]
http://actionscript-hindi.blogspot.com/2019/09/how-to-include-64-bit-and-32-bit-native.html

**Comment by P2b10** [comment]
@ajwfrost Thank´s a lot for your reply. I was doing something wrong on the atd.cfg file.

I was able to publish both apks (32 and 64) using a 1024 certificate in Animate 2015.2 with AIR SDK 33.1.1.86. Everything worked fine. Both apks are already on Google Console with no issues.

Best,

**Comment by ajwfrost** [comment]
@P2b10  great - at some point I still want to make all of this a bit simpler to manage..!

**Comment by roberto761** [comment]
Hello

I have been seeing all your comments and they are very interesting
I tell you my problems, which have to do with animate cc and HARMAN

1.- Creation of 32 bit and 64 bit files (to upload to google play)
My question is: which of the 2 options below is correct:
a) create 2 files: ARMv7 (32bit) and ARMv8 (64bit)
or
b) create 2 files: Intel x86 (32bit) and ARMv8 (64bit)

according to my logic, I did: option b)
so as not to cause update problems later, but I don't know if my decision is correct.

2.- I have seen that packing both apk (32 bit and 64 bit) does not work, my question is:

How to upload both apk files to Google play or what are the alternatives to upload correctly to Google play?


3.- how to edit the file correctly: adt.cfg
Should I edit the 2 lines to make them look like this?

DefaultArch=armv8
OverrideArch=armv8


I am a beginner in this and I appreciate your patience and can help me.

Greetings

**Comment by ajwfrost** [comment]
Hi @roberto761 

The Google Play rules mean that you can't just provide a 32-bit binary, you also (or instead) have to provide a 64-bit binary. But this is applied separately to the two CPU families, so if you provide x86 it would complain because you're not providing x64. To get the fullest support you'd want to have all 4 types - x86, x64, armv7, arvm8 - but most people only bother with the ARM-based CPUs I think.

You then need to adjust the version number slightly for each one because this generates multiple APKs and each APK must be distinguished. So if we take only the ARM binaries, then in order to support 32-bit devices you can set a version number on the 32-bit one of e.g. 1.2.3, but then when you generate the 64-bit APK it needs to have a higher version number such as 1.2.4, so that when the Play Store finds a device that's 64-bit, it knows which APK to supply (since a 64-bit device could run a 32-bit binary if it needed to..)

In terms of those cfg file values, it depends how you're packing things up. If you are using Animate, I would suggest commenting out those entries in the cfg file and instead using the UI settings within Animate, they should have all 4 CPU options available in the Publish Settings dialog now I believe? But if not, then my suggestion would be to keep the config values the same as each other and then change them each time
i.e.
1. set them to armv7, set your version number to e.g. 1.2.3, and build -> apk file should have the armv7 library in it
2. set them to armv8, set your version number to e.g. 1.2.4, and build -> apk file should have the armv8 library in it

thanks

**Comment by roberto761** [comment]
Hi ajwfrost

Thanks for your prompt response 

I give you more data, based on your comments

1.- I am developing - designing with in Animate cc 2020
2.- In the dialog box, I see 3 options
a) ARMv7 (32bit)
b) Intel x86 (32bit)
c) ARMv8 (64bits)

According to your comments I think that what I should do is this and I will appreciate if you tell me if I am making other mistakes:

1.- Create an apk (32 bits)
ARMv7 (32-bit)
example: supermarket (32bits) .apk
Version: 1.0.0

2.- Create an apk (64 bits)
ARMv8 (64-bit)
example: supermarket (64bits) .apk
Version: 1.0.1

3.- The adt.cfg file in its text contains these 2 lines:

# DefaultArch = armv8
# OverrideArch = armv8

should I write it like this?

DefaultArch = armv8
OverrideArch = armv8

and save the file?

Thanks again for taking the time to answer, your help is gold!

**Comment by ajwfrost** [comment]
Hi

If you're able to use Animate like that and set those values for ARMv7 and ARMv8, then you would need to comment out the lines from the adt.cfg file. The cfg file settings were added because we were changing things quicker than the Animate team could add the UI options, hence the "override" value would actually be used in preference to anything you set in their UI.

So if you just comment out (with the # symbol) those lines in adt.cfg, you just need to then do steps 1 and 2 to create your two APK files.

thanks

**Comment by roberto761** [comment]
Hi

excuse me

I did not understand your last comment:

if I create both files (ARMv7 32bits and ARMv8 64bits) with Animate CC 2020

what is the correct option in adt.cfg?

a) I put the symbol # and save the file
b) I remove the # symbol and save the file

Thank you for your patience .

**Comment by ajwfrost** [comment]
Yes, if you're changing it in Animate then use
```
#DefaultArch = armv8
#OverrideArch = armv8
```
in the config file and just change it via Animate

**Comment by roberto761** [comment]
ok

I will create both files

and the 2 lines of the adt.cfg file, will include the symbol #

Thanks

**Comment by roberto761** [comment]
hi ajwfrost

I have a question and I hope you can help me, without taking away much of your valuable time

With the new google policies, in August or September I understand that the new is:

change level 29
to use level 30

And although it is not yet the date for that change, I would like to know if you have had this type of topic to discuss.

I develop with animate CC, and it invites me to think about how the change will be (sdk, harman, animate cc version, etc)
you have already helped me in other previous doubts.

I look forward to your valuable comments

Greetings

**Reply by ajwfrost** [reply]
Hi

This sort of change is something we'd be able to update primarily within the SDK, but the slight challenge if you're using Animate or any other IDE is that these are set up currently to call the packaging tool for APK creation. When we move this to App Bundle creation, it may mean we need a workaround (such as we put in originally so that you could build an ARMv8 APK file, prior to the Animate adding that option to the UI).

I'll raise this with them, in case they're able to add an 'App Bundle' option. We still will likely need the APK creation so that you can actually test things locally; it would perhaps be better to create the APK from the Bundle though..

thanks

   Andrew

**Reply by roberto761** [reply]
Hi

I feel that maybe I did not explain it well

More than anything, my question refers to whether it will keep updating Harman's sdk (SDK versions), since for example, Google Play, constantly advances to the next API (I can do it at this time until number 29), but in August - September .... API 30 will be needed, and so on.

my fear is that ...
Let Google tell me when uploading my app:

"Your version (objective 29) is no longer allowed, use the latest one (30) ... and that no such API exists, taking into account that I develop with Animate CC"

that there will come a time when Google asks us for the latest APIs, and that when it cannot find them I can no longer upload my projects to google play.


I'm new to this and in fact thanks to your comments I was able to upload my first development to google play.

Thank you for your response, always constructive!
